src/mlang/m_ir/mir.ml

```diff
@@ -538,3 +543,18 @@ let fold_vars (f : Variable.t -> variable_data -> 'a -> 'a) (p : program) (acc :
           f var def acc)
         acc rule_data.rule_vars)
     p.program_rules acc
+
+let find_vars_by_io (p : program) (io_to_find : io) : VariableDict.t =
+  fold_vars
+    (fun var var_data acc ->
+      if
+        var_data.var_io = io_to_find
+        && var
+           = List.hd
+               (Pos.VarNameToID.find (Pos.unmark var.name) p.program_idmap
+               |> List.sort (fun v1 v2 ->
```

Comment:
OK that's kind of awkard but I see what you want to do. You want to remove the duplicates created by the phantom definitions of all variables to `undefined`. A better way to do this is simply to test whether the rule number of the execution number of the variable is `-1`. But maybe you also want to remove the duplicates of `VARTMP1` and only keep one instance?  Anyway, this behavior should be documented in `mir.mli`.